### PR TITLE
ARMeilleure: Hardware accelerate SHA256

### PR DESCRIPTION
**Pull request by merryhime**

### CODE REVIEW

File: ARMeilleure/Instructions/InstEmitSimdHashHelper.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+            String method = part2 ? nameof(SoftFallback.HashUpper) : nameof(SoftFallback.HashLower);
+            return context.Call(typeof(SoftFallback).GetMethod(method), x, y, w);
+        }
+
+        public static Operand EmitSha256su0(ArmEmitterContext context, Operand x, Operand y)
+        {
+            if (Optimizations.UseSha)
+            {
+                // return context.AddIntrinsic(Intrinsic.X86Sha256Msg1, x, y);
```
Comment (merryhime) [inline ARMeilleure/Instructions/InstEmitSimdHashHelper.cs:35]: Uncommenting this fails the Sha256su0_V test which confuses me immensely.

File: ARMeilleure/Instructions/InstEmitSimdHashHelper.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+            String method = part2 ? nameof(SoftFallback.HashUpper) : nameof(SoftFallback.HashLower);
+            return context.Call(typeof(SoftFallback).GetMethod(method), x, y, w);
+        }
+
+        public static Operand EmitSha256su0(ArmEmitterContext context, Operand x, Operand y)
+        {
+            if (Optimizations.UseSha)
+            {
+                return context.AddIntrinsic(Intrinsic.X86Sha256Msg1, context.Copy(x), y);
```
Comment (merryhime) [inline ARMeilleure/Instructions/InstEmitSimdHashHelper.cs:35]: I note not having the context.Copy here causes invalid register allocation?

Compare without the Copy (read from uninitalized register):
```asm
vmovdqu xmm0,XMMWORD PTR [rsp+0x4c]
vmovdqu xmm1,XMMWORD PTR [rsp+0x5c]
sha256msg1 xmm4,xmm1
vmovdqu xmm0,xmm4
vmovdqu XMMWORD PTR [rsp+0x4c],xmm0
```

To with the copy:
 ```asm
vmovdqu xmm4,XMMWORD PTR [rsp+0x4c]
vmovdqu xmm0,XMMWORD PTR [rsp+0x5c]
sha256msg1 xmm4,xmm0
vmovdqu xmm0,xmm4
vmovdqu XMMWORD PTR [rsp+0x4c],xmm0
```

Comment (gdkchan) [reply ARMeilleure/Instructions/InstEmitSimdHashHelper.cs:35]: It has to be handled on the pre-allocator, as iirc by default it does not expect the first operand and destination to share the same register, but I guess I was too slow and you already figured that out.

Comment (merryhime) [reply ARMeilleure/Instructions/InstEmitSimdHashHelper.cs:35]: Underlying issue found, see #3587.

File: ARMeilleure/CodeGen/X86/X86Instruction.cs [hunk 3]
```diff
@@ -180,6 +181,9 @@ enum X86Instruction
         Rsqrtss,
         Sar,
         Setcc,
+        Sha256Rnds2,
```
Comment (gdkchan) [inline ARMeilleure/CodeGen/X86/X86Instruction.cs:184]: Should be after `Sha256Msg2` for proper alphabetical sorting.

File: ARMeilleure/CodeGen/X86/IntrinsicTable.cs [hunk 4]
```diff
@@ -151,6 +152,9 @@ static IntrinsicTable()
             Add(Intrinsic.X86Roundss,      new IntrinsicInfo(X86Instruction.Roundss,      IntrinsicType.BinaryImm));
             Add(Intrinsic.X86Rsqrtps,      new IntrinsicInfo(X86Instruction.Rsqrtps,      IntrinsicType.Unary));
             Add(Intrinsic.X86Rsqrtss,      new IntrinsicInfo(X86Instruction.Rsqrtss,      IntrinsicType.Unary));
+            Add(Intrinsic.X86Sha256Rnds2,  new IntrinsicInfo(X86Instruction.Sha256Rnds2,  IntrinsicType.Ternary));
```
Comment (gdkchan) [inline ARMeilleure/CodeGen/X86/IntrinsicTable.cs:155]: Sorting.

File: ARMeilleure/CodeGen/X86/PreAllocator.cs [hunk 5]
```diff
@@ -309,10 +309,12 @@ void SplitOperand(Operand source, Operand lr, Operand hr)
                 case Instruction.Extended:
                 {
                     // BLENDVPD, BLENDVPS, PBLENDVB last operand is always implied to be XMM0 when VEX is not supported.
-                    if ((node.Intrinsic == Intrinsic.X86Blendvpd ||
-                         node.Intrinsic == Intrinsic.X86Blendvps ||
-                         node.Intrinsic == Intrinsic.X86Pblendvb) &&
-                         !HardwareCapabilities.SupportsVexEncoding)
+                    // SHA256RNDS2 always has an implied XMM0 as a last operand.
+                    if (((node.Intrinsic == Intrinsic.X86Blendvpd ||
```
Comment (gdkchan) [inline ARMeilleure/CodeGen/X86/PreAllocator.cs:313]: I think this would look better if you extracted the sub-expression into a local variable like this:
```cs
bool isBlend = node.Intrinsic == Intrinsic.X86Blendvpd ||
               node.Intrinsic == Intrinsic.X86Blendvps ||
               node.Intrinsic == Intrinsic.X86Pblendvb;

if ((isBlend && !HardwareCapabilities.SupportsVexEncoding) || node.Intrinsic == Intrinsic.X86Sha256Rnds2)
```

File: ARMeilleure/Translation/PTC/Ptc.cs [hunk 6]
```diff
@@ -946,9 +947,13 @@ internal static bool GetEndianness()
             return BitConverter.IsLittleEndian;
         }
 
-        private static ulong GetFeatureInfo()
+        private static uint[] GetFeatureInfo()
         {
-            return (ulong)HardwareCapabilities.FeatureInfoEdx << 32 | (uint)HardwareCapabilities.FeatureInfoEcx;
+            return new uint[] {
```
Comment (gdkchan) [inline ARMeilleure/Translation/PTC/Ptc.cs:952]: `{` on new line.

File: ARMeilleure/Translation/PTC/Ptc.cs [hunk 7]
```diff
@@ -997,6 +1002,17 @@ public bool IsHeaderValid()
 
                 return XXHash128.ComputeHash(MemoryMarshal.AsBytes(spanHeader).Slice(0, Unsafe.SizeOf<OuterHeader>() - Unsafe.SizeOf<Hash128>())) == HeaderHash;
             }
+
+            public uint[] FeatureInfo
+            {
+                get { return new uint[3] { FeatureInfo0, FeatureInfo1, FeatureInfo2 }; }
```
Comment (gdkchan) [inline ARMeilleure/Translation/PTC/Ptc.cs:1008]:
```suggestion
                get => new uint[3] { FeatureInfo0, FeatureInfo1, FeatureInfo2 };
```

File: ARMeilleure/Translation/PTC/Ptc.cs [hunk 8]
```diff
@@ -24,10 +25,10 @@ namespace ARMeilleure.Translation.PTC
 {
     public static class Ptc
     {
-        private const string OuterHeaderMagicString = "PTCohd\0\0";
+        private const string OuterHeaderMagicString = "PTCohd\x01\0";
```
Comment (gdkchan) [inline ARMeilleure/Translation/PTC/Ptc.cs:28]: Why is that needed? From a quick glance at the code, it seems that just the InternalVersion change would be enough, but maybe I missed something.

File: ARMeilleure/Instructions/InstEmitSimdHash.cs [hunk 9]
```diff
@@ -3,6 +3,7 @@
 using ARMeilleure.Translation;
 
 using static ARMeilleure.Instructions.InstEmitHelper;
+using static ARMeilleure.IntermediateRepresentation.Operand.Factory;
```
Comment (gdkchan) [inline ARMeilleure/Instructions/InstEmitSimdHash.cs:6]: Unused?

**Review by gdkchan — COMMENTED** [review]
I think it should be merged after #3587 to avoid the copy workaround.

File: ARMeilleure/Translation/PTC/Ptc.cs (hunk 8 above)
Comment (merryhime) [reply ARMeilleure/Translation/PTC/Ptc.cs:28]: Location of fields in OuterHeader have shifted; I'm reasonably happy to ignore this if you want.

**Review by gdkchan — APPROVED** [review]
lgtm, thanks.

File: ARMeilleure/Translation/PTC/Ptc.cs [hunk 10]
```diff
@@ -234,7 +235,7 @@ private static unsafe bool Load(string fileName, bool isBackup)
                     return false;
                 }
 
-                if (outerHeader.FeatureInfo != GetFeatureInfo())
+                if (!outerHeader.FeatureInfo.SequenceEqual(GetFeatureInfo()))
```
Comment (LDj3SNuD) [inline ARMeilleure/Translation/PTC/Ptc.cs:238]:
```suggestion
                if (outerHeader.FeatureInfo != GetFeatureInfo())
```

File: ARMeilleure/Translation/PTC/Ptc.cs [hunk 11]
```diff
@@ -13,6 +13,7 @@
 using System.Diagnostics;
 using System.IO;
 using System.IO.Compression;
+using System.Linq;
```
Comment (LDj3SNuD) [inline ARMeilleure/Translation/PTC/Ptc.cs:16]:
```suggestion
```

File: ARMeilleure/Translation/PTC/Ptc.cs [hunk 12]
```diff
@@ -946,9 +947,14 @@ internal static bool GetEndianness()
...
-        private static ulong GetFeatureInfo()
+        private static uint[] GetFeatureInfo()
         {
-            return (ulong)HardwareCapabilities.FeatureInfoEdx << 32 | (uint)HardwareCapabilities.FeatureInfoEcx;
+            return new uint[]
+            {
+                (uint)HardwareCapabilities.FeatureInfo1Ecx,
+                (uint)HardwareCapabilities.FeatureInfo1Edx,
+                (uint)HardwareCapabilities.FeatureInfo7Ebx
+            };
```
Comment (LDj3SNuD) [inline ARMeilleure/Translation/PTC/Ptc.cs:957]:
```suggestion
            return new FeatureInfo(
                (uint)HardwareCapabilities.FeatureInfo1Ecx,
                (uint)HardwareCapabilities.FeatureInfo1Edx,
                (uint)HardwareCapabilities.FeatureInfo7Ebx);
```

File: ARMeilleure/Translation/PTC/Ptc.cs [hunk 13]
```diff
@@ -946,9 +947,14 @@ internal static bool GetEndianness()
             return BitConverter.IsLittleEndian;
         }
 
-        private static ulong GetFeatureInfo()
+        private static uint[] GetFeatureInfo()
```
Comment (LDj3SNuD) [inline ARMeilleure/Translation/PTC/Ptc.cs:950]:
```suggestion
        private static FeatureInfo GetFeatureInfo()
```

File: ARMeilleure/Translation/PTC/Ptc.cs [hunk 14]
```diff
@@ -968,15 +974,15 @@ private static uint GetOSPlatform()
...
+        [StructLayout(LayoutKind.Sequential, Pack = 1/*, Size = 54*/)]
         private struct OuterHeader
         {
             public ulong Magic;
 
             public uint CacheFileVersion;
 
             public bool Endianness;
-            public ulong FeatureInfo;
+            public uint FeatureInfo0, FeatureInfo1, FeatureInfo2;
```
Comment (LDj3SNuD) [inline ARMeilleure/Translation/PTC/Ptc.cs:985]:
```suggestion
            public FeatureInfo FeatureInfo;
```

File: ARMeilleure/Translation/PTC/Ptc.cs [hunk 15]
```diff
@@ -997,6 +1003,17 @@ public bool IsHeaderValid()
...
+            public uint[] FeatureInfo
+            {
+                get => new uint[3] { FeatureInfo0, FeatureInfo1, FeatureInfo2 };
+                set
+                {
+                    FeatureInfo0 = value[0];
+                    FeatureInfo1 = value[1];
+                    FeatureInfo2 = value[2];
+                }
+            }
```
Comment (LDj3SNuD) [inline ARMeilleure/Translation/PTC/Ptc.cs:1016]:
```suggestion
```

Comment (LDj3SNuD) [reply ARMeilleure/Translation/PTC/Ptc.cs:1016]: And finally add this record struct:
```csharp
        [StructLayout(LayoutKind.Sequential, Pack = 1/*, Size = 12*/)]
        private record struct FeatureInfo(uint FeatureInfo0, uint FeatureInfo1, uint FeatureInfo2);
```
below the `OuterHeader` struct.

**Review by LDj3SNuD — COMMENTED** [review]
Apart from the recommended code, lgtm; I tested locally the slow paths but not the fast paths (not having a new cpu).

**Review by LDj3SNuD — APPROVED** [review]
